Fix bounds for sympy expressions

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102611
* __->__ #102722

The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225

lezcano committed:
Fix bounds for sympy expressions

The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

[ghstack-poisoned]

lezcano commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -844,83 +844,6 @@ def eval(cls, base, divisor):
         else:
             return base / divisor
 
-class FloorDiv(sympy.Function):
```
into `_sympy/functions.py`

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

ezyang commented on torch/_inductor/optimize_indexing.py:
```diff
@@ -63,74 +62,6 @@ def range_expressable_in_32_bits(range):
     )
 
 
-def get_expr_range(expr, vars_ranges: dict):
```
Did this get moved?

lezcano replied:
This one was incorrect and it's now replaced by `bound_sympy`.

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -19,15 +22,15 @@ class ValueRangeError(RuntimeError):
-    if isinstance(e, int):
+    if isinstance(e, bool):
+        return sympy.true if e else sympy.false
+    elif isinstance(e, int):
```
nice catch

ezyang commented on torch/utils/_sympy/interp.py:
```diff
@@ -37,7 +40,7 @@ def handlers():
         sympy.Not: "not_",
         TrueDiv: "truediv",
-        FloorDiv: "div",
+        FloorDiv: "floordiv",
```
Is this OK? I was under the impression the analysis was also used by Inductor, which has its own naming convention for operation methods we need to abide by.

lezcano replied:
I think this is the right one, yes. See https://github.com/pytorch/pytorch/pull/102722#discussion_r1214387734

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -251,59 +219,23 @@ def ge(cls, a, b):
-        return ValueRanges.decreasing_map(x, operator.neg)
+    @classmethod
+    def div(cls, a, b):
+        return cls.truediv(a, b)
```
Once again, I'm not sure this is kosher because it is inconsistent with Inductor cpp codegen

```
    @staticmethod
    def div(a, b):
        return f"{a} / {b}"
```

from codegen.cpp

lezcano replied:
Note that this change was not introduced in this PR, but in a previous one. The relevant discussion is in https://github.com/pytorch/pytorch/pull/100547#discussion_r1184646605

ezyang commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -444,21 +360,114 @@ def fn_(x, y):
+        x = ValueRanges.wrap(x)
+        if x.lower < 0:
+            return ValueRanges.unknown()
+        return ValueRanges.increasing_map(x, sympy.sqrt)
```
Which of these changed, which are the same?

lezcano replied:
Yeah, the diff is just terrible. Let me split the "moving things around" and the "improve things" into a few PRs

ezyang reviewed (approved):
But please see comments about div/floordiv confusion

lezcano commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -147,66 +152,31 @@ def coordinatewise_monotone_map(cls, x, y, fn):
-        return index
+    @staticmethod
+    def not_(a):
+        return ValueRanges.decreasing_map(a, sympy.Not)
```
Sure, I can do that.

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -3548,13 +3549,13 @@ def fn(idx):
-        phend = ops.index_expr(ir.FloorDiv(h, stride[0]) + 1, torch.int32)
-        pwend = ops.index_expr(ir.FloorDiv(w, stride[1]) + 1, torch.int32)
+        phend = ops.index_expr(FloorDiv(h, stride[0]) + 1, torch.int32)
+        pwend = ops.index_expr(FloorDiv(w, stride[1]) + 1, torch.int32)
```
I just needed to move all the relevant special classes into one folder so that I can import them properly in the other parts of the analysis. All the changers in this file just remove the `ir.` from the import.
In all fairness, I do think it's nicer without the `ir.`, as this is not an IR class, but a SymPy construct.

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

lezcano committed:
Update on "Fix bounds for sympy expressions"


The analysis for SymPy expressions was incorrect as, even though it said
that the assumption was "smoothness" the assumtion was, in fact, that he
formula was monotone in every variable. In other words, it was
assuming that the derivative does not change signs in any variable (!!).

These issues showed up when using this analysis on a larger set of
formulas, where we would get bounds where lower > upper when using
ModularIndexing.

This PR extends the bounds analysis to SymPy expressions. It also
separates the symbols occuring on SymPy expressions into their own
class, as we don't want to use any fallback in this case.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225

[ghstack-poisoned]

lezcano commented:
split into https://github.com/pytorch/pytorch/pull/104556 https://github.com/pytorch/pytorch/pull/104557 https://github.com/pytorch/pytorch/pull/104559
